### ISSUE TITLE
Add new check for error ignoring on command substitutions

### ISSUE THREAD

**Issue by felipecrs** [koalaman/shellcheck#2338]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Suggestion

Given the following script:

```sh
#!/bin/bash

set -e

basename "$(curl -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
```

It would normally work, but `curl` could actually fail (such as with command not found). So, for the sake of the example, I will force it to fail by using `curl2`.

Examples:

```bash
$ bash <<'EOF'; echo $?
#!/bin/bash

set -e

basename "$(curl -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
EOF
v2.0.0
0
```

With `curl2`:

```bash
$ bash <<'EOF'; echo $?
#!/bin/bash

set -e

basename "$(curl2 -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
EOF
bash: line 5: curl2: command not found

0
```

The status code of the latter example is 0 even if the command failed. This is misleading. ShellCheck could advise to be written like so instead:

```bash
$ bash <<'EOF'; echo $?
#!/bin/bash

set -e

curl_output="$(curl2 -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
basename "$curl_output"
EOF
bash: line 5: curl2: command not found
127
```

References https://github.com/twpayne/chezmoi/issues/1466#issuecomment-929709757

**Comment by kurahaupo** [comment]
Yet another strike against `set -e` for creating expectations that directly conflict with how the shell actually works and how correct, idiomatic shell code is normally written.

If such a warning were given, I would like it to be gated on the presence of `set -e`, and for the wording of the warning to point out that `set -e` makes promises that cannot be met.

**Comment by felipecrs** [comment]
> If such a warning were given, I would like it to be gated on the presence of `set -e`, and for the wording of the warning to point out that `set -e` makes promises that cannot be met.

I definitely agree with this.

**Comment by koalaman** [comment]
Thanks to @DoxasticFox you can now enable an optional check for this:

```
#!/bin/bash
# shellcheck enable=check-extra-masked-returns
set -e
basename "$(curl -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
```

results in

```
basename "$(curl -fsSL -w '%{url_effective}' -o /dev/null https://github.com/docker/compose/releases/latest)"
            ^-- SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).
```

**Comment by felipecrs** [comment]
This is awesome!

**Comment by felipecrs** [comment]
Using this feature, I realized there are:

```bash
[[ $(command -v curl) ]]
```

and

```bash
my_function "$(cat <<EOF
my
multiline
parameter
EOF
)"
```

Which are two cases very known, and return code on them is very likely not to mean anything. Should this rule have exceptions like these above?

**Comment by felipecrs** [comment]
That's right. Thanks for the rationale.

Perhaps an alternative solution would be to allow `# shellcheck disable=check-extra-masked-returns` instead of having to use `|| true`, as it would be just a comment rather than being an actual command in the code.

**Comment by GhostLyrics** [comment]
Is enabling this optional rule necessary if you run your (bash) script with `-o pipefail`? I think the [wiki page for this check](https://github.com/koalaman/shellcheck/wiki/SC2312) should contain a hint about whether this makes sense.

**Comment by felipecrs** [comment]
I don't get it. What is the relationship between this and -o pipefail?

**Comment by GhostLyrics** [comment]
> I don't get it. What is the relationship between this and `-o pipefail`?

For context - I run my scripts with `set -euo pipefail`.

For example, I scanned my existing code:

## example 1
```text
In resources/scripts/helper/REDACTED_lib.sh line 269:
            unprocessed_jobs_count=$(echo "$unprocessed_jobs" | tail --lines +2 | wc --lines)
                                                                ^-------------^ SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).
```

In my understanding enabling pipefail will _not_ cause the errors to be swallowed because the entire "pipeline" will fail as soon as the first step fails.

## example 2
Similarly here - if one step fails, the pipeline will fail and it would use the `else` branch of my code:

```text
In resources/scripts/helper/REDACTED_lib.sh line 275:
        if ! jobs_in_queue=$(REDACTED status | tail --lines +2 | grep --count --invert-match "$excluded_strings"); then
                             ^-------------^ SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).
                                               ^-------------^ SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).
```

Scanned section in question for example 2:
```bash
# grep returns code 0 with no results
if ! jobs_in_queue=$(REDACTED status | tail --lines +2 | grep --count --invert-match "$excluded_strings"); then
    if [ "$jobs_in_queue" == 0 ]; then
        REDACTED "Processed all pending jobs."
        break
    fi

    REDACTED "REDACTED: could not retrieve pending jobs." "$jobs_in_queue"
fi
```

**Comment by DoxasticFox** [comment]
@GhostLyrics `SC2312` will be suppressed in pipelines if you use `set -o pipefail`, but there's other ways a command's exit code can be masked. The [unit tests](https://github.com/DoxasticFox/shellcheck/blob/093df8cb2448a3b2f352b4e3105d2af45ccf3206/src/ShellCheck/Analytics.hs#L4744-L4771) give examples.

**Comment by kurahaupo** [comment]
> ```shell
> [[ $(command -v curl) ]]
> ```

This is sadly a badly over-used idiom.

1. In many cases (including `command -v`) the exit status will tell you what you want to know, and the output can be entirely ignored: `command -v curl >/dev/null 2>&1`
1. If the objective is truly to test whether the output of a command is not empty, it's not necessary to wait for the entire output of the command to finish. Once it outputs a single byte, you have the answer to that question: `somecmd | read` will tell you whether the output of `somecmd` is non-empty, without waiting for `somecmd` to finish. (Normally somecmd would be terminated almost immediately with `SIGPIPE`, and that would be a good thing.)

> ```shell
> my_function "$(cat <<EOF
> my
> multiline
> parameter
> EOF
> )"
> ```

Multiline strings are valid, so if it's simple just use:
```lang=shell
my_function 'my
multiline
parameter'
```
If you have complex embedded quoting, it would probably make the code clearer if the value is in a variable anyway:
```lang=shell
sql_expression=$(cat) <<\EOF
SELECT count(*) as `Num`,
       'String with "Quotes"' as `Label`
FROM bananas
WHERE ripeness like '%firm%'
AND price = '$14'
EOF
my_sql_cmd -e "$sql_expression"
```

Either way, the relevant exit status is more obvious to ShellCheck.
